src/viam/app/data/client.py

```diff
@@ -290,10 +328,226 @@ async def bounding_box_labels_by_filter(self, filter: Optional[Filter] = None) -
+        metadata.
+
+        Raises:
+            GRPCError: If an invalid part ID is passed.
+            AssertionError: If a list of Timestamps is provided and its length does not match the length of the list of tabular data.
+        """
+        sensor_contents = [None] * len(tabular_data)
+        if timestamps:
+            assert len(timestamps) == len(tabular_data)
+
+        for i in range(len(tabular_data)):
+            s = Struct()
+            s.update(tabular_data[i])
+            sensor_contents[i] = SensorData(
+                metadata=SensorMetadata(
+                    time_requested=timestamps[i][0] if timestamps and timestamps[i][0] else None,
+                    time_received=timestamps[i][1] if timestamps and timestamps[i][1] else None
+                ),
+                struct=s
+            )
+
+        metadata = UploadMetadata(
+            part_id=part_id,
+            component_type=component_type,
+            component_name=component_name,
+            method_name=method_name,
+            type=DataType.DATA_TYPE_TABULAR_SENSOR,
+            file_name=None,  # Not used in app.
+            method_parameters=method_parameters if method_parameters else None,
+            file_extension=file_extension if file_extension else None,
+            tags=tags if tags else None,
+        )
+        _: DataCaptureUploadResponse = await self._data_capture_upload(metadata=metadata, sensor_contents=sensor_contents)
+
+    async def _data_capture_upload(self, metadata: UploadMetadata, sensor_contents: List[SensorData]) -> DataCaptureUploadResponse:
+        request = DataCaptureUploadRequest(metadata=metadata, sensor_contents=sensor_contents)
+        response: DataCaptureUploadResponse = await self._data_sync_client.DataCaptureUpload(request, metadata=self._metadata)
+        return response
+
+    async def file_upload(
+        self,
+        part_id: str,
+        component_type: Optional[str],
+        component_name: Optional[str],
+        method_name: Optional[str],
+        file_name: Optional[str],
+        method_parameters: Optional[Mapping[str, Any]],
+        file_extension: Optional[str],
+        tags: Optional[List[str]],
+        data: Optional[bytes]
+    ) -> None:
+        """Upload arbitrary file data.
+
+        Sync file data that may be stored on a robot along with the relevant metadata to app.viam.com.
+
+        Args:
+            part_id (str): Part ID of the component associated with the file.
+            component_type (Optional[str]): Optional type of the component associated with the file (i.e., "movement_sensor").
+            component_name (Optional[str]): Optional name of the component associated with the file.
+            method_name (Optional[str]): Optional name of the method associated with the file.
+            file_name (Optional[str]): Optional name of the file.
```

Comment:
Naively, I would think that a method named `file_upload` would require a file name/extension. What happens if they aren't provided? Can we clarify in the comments?

Reply:
If they are not provided, the data still gets uploaded and the name and extension fields will just use the empty string "" as their values.